File: backend/app/schemas/finance.py
"""Finance Intelligence API response schemas."""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.common import ResponseMeta, utc_now


class FinanceMetric(BaseModel):
    id: str
    label: str
    value: float | int | None = None
    formatted_value: str | None = None
    unit: str | None = None
    delta: float | None = None
    delta_label: str | None = None
    trend: Literal["up", "down", "flat"] | None = None
    format: Literal["currency", "percent", "number"] | None = None
    available: bool = True
    source: str | None = None


class FinanceOverviewResponse(BaseModel):
    metrics: list[FinanceMetric]
    period_start: date | None = None
    period_end: date | None = None
    generated_at: datetime = Field(default_factory=utc_now)
    meta: ResponseMeta = Field(default_factory=ResponseMeta)


class ProfitabilityRow(BaseModel):
    region: str
    revenue: float | None = None
    cost: float | None = None
    profit: float | None = None
    margin: float | None = None
    growth: float | None = None


class ProfitabilityResponse(BaseModel):
    rows: list[ProfitabilityRow]
    available: bool = True
    meta: ResponseMeta = Field(default_factory=ResponseMeta)


class CostBreakdownRow(BaseModel):
    cost_category: str
    amount: float | None = None
    percentage: float | None = None
    trend: Literal["up", "down", "flat"] | None = None


class CostBreakdownResponse(BaseModel):
    rows: list[CostBreakdownRow]
    available: bool = True
    meta: ResponseMeta = Field(default_factory=ResponseMeta)


class CashflowRow(BaseModel):
    period: str
    inflows: float | None = None
    outflows: float | None = None
    net_cashflow: float | None = None
    profit: float | None = None
    margin: float | None = None


class CashflowResponse(BaseModel):
    rows: list[CashflowRow]
    available: bool = True
    meta: ResponseMeta = Field(default_factory=ResponseMeta)


class FinancialRiskRow(BaseModel):
    risk: str
    severity: Literal["low", "medium", "high", "critical"] = "medium"
    estimated_impact: float | None = None
    owner: str | None = None
    recommendation: str | None = None


class FinancialRisksResponse(BaseModel):
    rows: list[FinancialRiskRow]
    available: bool = True
    meta: ResponseMeta = Field(default_factory=ResponseMeta)


class BudgetVarianceRow(BaseModel):
    department: str
    budget: float | None = None
    actual: float | None = None
    variance: float | None = None
    variance_pct: float | None = None


class BudgetVarianceResponse(BaseModel):
    rows: list[BudgetVarianceRow]
    available: bool = True
    source: str | None = "campaign_performance"
    meta: ResponseMeta = Field(default_factory=ResponseMeta)

File: frontend/src/services/finance.ts
/**
 * Finance Intelligence API client.
 */

import { apiFetch } from "@/services/api-client";

export type FinanceMetric = {
  id: string;
  label: string;
  value: number | null;
  formatted_value: string | null;
  unit: string | null;
  delta: number | null;
  delta_label: string | null;
  trend: "up" | "down" | "flat" | null;
  format: "currency" | "percent" | "number" | null;
  available: boolean;
  source: string | null;
};

export type FinanceOverview = {
  metrics: FinanceMetric[];
  period_start: string | null;
  period_end: string | null;
  generated_at: string;
};

export type ProfitabilityRow = {
  region: string;
  revenue: number | null;
  cost: number | null;
  profit: number | null;
  margin: number | null;
  growth: number | null;
};

export type Profitability = {
  rows: ProfitabilityRow[];
  available: boolean;
};

export type CostBreakdownRow = {
  cost_category: string;
  amount: number | null;
  percentage: number | null;
  trend: "up" | "down" | "flat" | null;
};

export type CostBreakdown = {
  rows: CostBreakdownRow[];
  available: boolean;
};

export type CashflowRow = {
  period: string;
  inflows: number | null;
  outflows: number | null;
  net_cashflow: number | null;
  profit: number | null;
  margin: number | null;
};

export type Cashflow = {
  rows: CashflowRow[];
  available: boolean;
};

export type FinancialRiskRow = {
  risk: string;
  severity: "low" | "medium" | "high" | "critical";
  estimated_impact: number | null;
  owner: string | null;
  recommendation: string | null;
};

export type FinancialRisks = {
  rows: FinancialRiskRow[];
  available: boolean;
};

export type BudgetVarianceRow = {
  department: string;
  budget: number | null;
  actual: number | null;
  variance: number | null;
  variance_pct: number | null;
};

export type BudgetVariance = {
  rows: BudgetVarianceRow[];
  available: boolean;
};

export type FinanceQueryParams = {
  dateFrom?: string;
  dateTo?: string;
  regionIds?: string[];
  departmentIds?: string[];
  costCategoryIds?: string[];
  search?: string;
};

function toQuery(params: Record<string, string | number | undefined | null>): string {
  const sp = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value === undefined || value === null || value === "") continue;
    sp.set(key, String(value));
  }
  const q = sp.toString();
  return q ? `?${q}` : "";
}

function filterParams(filters: FinanceQueryParams) {
  return {
    date_from: filters.dateFrom || undefined,
    date_to: filters.dateTo || undefined,
    region: filters.regionIds?.length ? filters.regionIds.join(",") : undefined,
    department: filters.departmentIds?.length
      ? filters.departmentIds.join(",")
      : undefined,
    cost_category: filters.costCategoryIds?.length
      ? filters.costCategoryIds.join(",")
      : undefined,
    search: filters.search || undefined,
  };
}

const BASE = "/api/v1/finance";

export const financeApi = {
  overview: (filters: FinanceQueryParams = {}) =>
    apiFetch<FinanceOverview>(`${BASE}/overview${toQuery(filterParams(filters))}`),
  profitability: (filters: FinanceQueryParams = {}) =>
    apiFetch<Profitability>(`${BASE}/profitability${toQuery(filterParams(filters))}`),
  costBreakdown: (filters: FinanceQueryParams = {}) =>
    apiFetch<CostBreakdown>(`${BASE}/cost-breakdown${toQuery(filterParams(filters))}`),
  cashflow: (filters: FinanceQueryParams = {}) =>
    apiFetch<Cashflow>(`${BASE}/cashflow${toQuery(filterParams(filters))}`),
  financialRisks: (filters: FinanceQueryParams = {}) =>
    apiFetch<FinancialRisks>(
      `${BASE}/financial-risks${toQuery(filterParams(filters))}`,
    ),
  budgetVariance: (filters: FinanceQueryParams = {}) =>
    apiFetch<BudgetVariance>(
      `${BASE}/budget-variance${toQuery(filterParams(filters))}`,
    ),
};

export const financeQueryKeys = {
  all: ["finance"] as const,
  overview: (filters: FinanceQueryParams) =>
    [...financeQueryKeys.all, "overview", filters] as const,
  profitability: (filters: FinanceQueryParams) =>
    [...financeQueryKeys.all, "profitability", filters] as const,
  costs: (filters: FinanceQueryParams) =>
    [...financeQueryKeys.all, "costs", filters] as const,
  cashflow: (filters: FinanceQueryParams) =>
    [...financeQueryKeys.all, "cashflow", filters] as const,
  risks: (filters: FinanceQueryParams) =>
    [...financeQueryKeys.all, "risks", filters] as const,
  budget: (filters: FinanceQueryParams) =>
    [...financeQueryKeys.all, "budget", filters] as const,
};
